### DISCUSSION TITLE
Having issues using filters

### DISCUSSION THREAD

**Discussion by tonislcs** [supabase-community/supabase-csharp#142] [Q&A]
Having issues using filters, tried looking at the documentation but i still couldn't figure out by what it means `Advanced filters require a constructor with more specific arguments`

`Model [Table("servers")]
public class Servers : BaseModel
{
    [PrimaryKey]
    [Column("id")]
    public float id { get; set; }
    
    [Column("pterodactylid")]
    public string PterodactylId { get; set; }
    
    [Column("ip")]
    public string Ip { get; set; }
    
    [Column("port")]
    public string? Port { get; set; }
    
    [Column("players")]
    public float? Players { get; set; }
    
    [Column("hostname")]
    public string? Hostname { get; set; }

    [Column("playerlist")]
    public string[]? Playerlist { get; set; }
    
    [Column("category")]
    public string Category { get; set; }
    
    [Column("upanelid")]
    public string Upanelid { get; set; }
    
}`

and this is the query am trying to make `    public async Task<float> FindServerByPlayer(string player)
    {
        var response = await _supabase
            .From<Servers>()
            .Select("id")
            .Where(x => x.Playerlist, Postgrest.Constants.Operator.Contains, player)
            .Get();
        Console.WriteLine(response.Content);
        var jsonDocument = JsonDocument.Parse(response.Content);
        var rootElement = jsonDocument.RootElement;

        return rootElement[0].GetProperty("id").GetSingle();
    }`

**Comment by acupofjose** [comment]
It looks like you should probably be using the `IN` or `LIKE` operator rather than `CONTAINS` - could you try changing operators and see if that fixes your problem?

**Comment by tonislcs** [comment]
For e xample if i use "LIKE" I get this `{"code":"42883","details":null,"hint":"No operator matches the given name and argument types. You might need to add explicit type casts.","message":"operator does not exist: text[] ~~ unknown"}
22
` and for IN same error Advanced filter

**Reply by acupofjose** [reply]
Okay, one more. And if this works then it we'll just need to open an issue on clarifying documentation.

Try changing to:
```c#
var response = await _supabase
.From()
.Select("id")
.Where(x => x.Playerlist, Operator.In, new List<string> { player })
.Get();
```

**Reply by neuralworm** [reply]
This doesn't work for me, as the Where method expects a bool as the intended return type of the first argument.  

I've been facing the same issue, and whether it's a string, list, or array, the Filter method seems unable to recognize the provided criterion as valid,

**Comment by neuralworm** [comment]
> Having issues using filters, tried looking at the documentation but i still couldn't figure out by what it means `Advanced filters require a constructor with more specific arguments`
> 
> `Model [Table("servers")] public class Servers : BaseModel { [PrimaryKey] [Column("id")] public float id { get; set; }
> 
> ```
> [Column("pterodactylid")]
> public string PterodactylId { get; set; }
> 
> [Column("ip")]
> public string Ip { get; set; }
> 
> [Column("port")]
> public string? Port { get; set; }
> 
> [Column("players")]
> public float? Players { get; set; }
> 
> [Column("hostname")]
> public string? Hostname { get; set; }
> 
> [Column("playerlist")]
> public string[]? Playerlist { get; set; }
> 
> [Column("category")]
> public string Category { get; set; }
> 
> [Column("upanelid")]
> public string Upanelid { get; set; }
> ```
> 
> }`
> 
> and this is the query am trying to make ` public async Task FindServerByPlayer(string player) { var response = await _supabase .From() .Select("id") .Where(x => x.Playerlist, Postgrest.Constants.Operator.Contains, player) .Get(); Console.WriteLine(response.Content); var jsonDocument = JsonDocument.Parse(response.Content); var rootElement = jsonDocument.RootElement;
> 
> ```
>     return rootElement[0].GetProperty("id").GetSingle();
> }`
> ```

From what I see, it appears your trying to find a row where the column being filtered is a String Array, and its being searched for in a single string "Player", and so it can't search for an array inside of a string.  But perhaps I misunderstand how the Filter method works.  I've been facing a similar issue with the following code.

`var res = await _supabase._client.From<SteamMarketPrice>().Select("*").Filter(x => x.name, Postgrest.Constants.Operator.In, hashNames).Get();`

Where x.name is a string, and the variable hashNames is a List of strings to search within.

**Comment by neuralworm** [comment]
I pulled up the inner workings of the FIlter method [here](https://github.com/supabase-community/postgrest-csharp/blob/9f9f72dbf43e7bcc1546b8abd6207b48e190e6e1/Postgrest/Table.cs#L102) as inspecting this may help:
` 

        public Table<TModel> Filter<TCriterion>(Expression<Func<TModel, object>> predicate, Operator op,
            TCriterion? criterion)
        {
            var visitor = new SelectExpressionVisitor();
            visitor.Visit(predicate);

            if (visitor.Columns.Count == 0)
                throw new ArgumentException("Expected predicate to return a reference to a Model column.");

            if (visitor.Columns.Count > 1)
                throw new ArgumentException("Only one column should be returned from the predicate.");

            return Filter(visitor.Columns.First(), op, criterion);
        }

        /// <inheritdoc />
        public Table<TModel> Filter<TCriterion>(string columnName, Operator op, TCriterion? criterion)
        {
            switch (criterion)
            {
                case null:
                    switch (op)
                    {
                        case Operator.Equals:
                        case Operator.Is:
                            _filters.Add(new QueryFilter(columnName, Operator.Is, QueryFilter.NullVal));
                            break;
                        case Operator.Not:
                        case Operator.NotEqual:
                            _filters.Add(new QueryFilter(columnName, Operator.Not,
                                new QueryFilter(columnName, Operator.Is, QueryFilter.NullVal)));
                            break;
                        default:
                            throw new PostgrestException(
                                    "NOT filters must use the `Equals`, `Is`, `Not` or `NotEqual` operators")
                                { Reason = FailureHint.Reason.InvalidArgument };
                    }

                    return this;
                case string stringCriterion:
                    _filters.Add(new QueryFilter(columnName, op, stringCriterion));
                    return this;
                case int intCriterion:
                    _filters.Add(new QueryFilter(columnName, op, intCriterion));
                    return this;
                case float floatCriterion:
                    _filters.Add(new QueryFilter(columnName, op, floatCriterion));
                    return this;
                case List<object> listCriteria:
                    _filters.Add(new QueryFilter(columnName, op, listCriteria));
                    return this;
                case Dictionary<string, object> dictCriteria:
                    _filters.Add(new QueryFilter(columnName, op, dictCriteria));
                    return this;
                case IntRange rangeCriteria:
                    _filters.Add(new QueryFilter(columnName, op, rangeCriteria));
                    return this;
                case FullTextSearchConfig fullTextSearchCriteria:
                    _filters.Add(new QueryFilter(columnName, op, fullTextSearchCriteria));
                    return this;
                default:
                    throw new PostgrestException(
                        "Unknown criterion type, is it of type `string`, `int`, `float`, `List`, `Dictionary<string, object>`, `FullTextSearchConfig`, or `Range`?")
                    {
                        Reason = FailureHint.Reason.InvalidArgument
                    };
            }
        }`

**Comment by neuralworm** [comment]
Well, I finally figured out my issue which got me the warning `Unknown criterion type, is it of type "string","int", "float", "List", "Dictionary<string, object>", "FullTextSearchConfig", or "Range"?`.  The FIlter method would not take a preconstructed string list of type `List<string>` as its TCriterion.  I needed to construct a new List with the exact type specified as `List<object>`, like so:  `.Filter(x => x.hash_name, Postgrest.Constants.Operator.In, new List<object>(hashNames))` where `hashNames` is of type `List<string>`.

This may sound silly, but maybe try something like `new string(player)` as your criterion?

**Reply by acupofjose** [reply]
Hm. This seems a bit like a regression while we were trying to provide better type casting on query filters. Glad you got this figured out! But I'm tracking the suggestions/issue here: https://github.com/supabase-community/postgrest-csharp/issues/87
